Src/IronPython/Modules/Builtin.cs

```diff
@@ -325,26 +325,45 @@ public static partial class Builtin {
         }
 
         [LightThrowing]
-        public static object eval(CodeContext/*!*/ context, string expression) => eval(context, expression, null, null);
+        public static object eval(CodeContext/*!*/ context, object expression) => eval(context, expression, null, null);
 
         [LightThrowing]
-        public static object eval(CodeContext/*!*/ context, string expression, PythonDictionary globals) => eval(context, expression, globals, null);
+        public static object eval(CodeContext/*!*/ context, object expression, PythonDictionary globals) => eval(context, expression, globals, null);
 
         [LightThrowing]
-        public static object eval(CodeContext/*!*/ context, string expression, PythonDictionary globals, object locals) {
+        public static object eval(CodeContext/*!*/ context, object expression, PythonDictionary globals, object locals) {
             Debug.Assert(context != null);
-            if (expression == null) throw PythonOps.TypeError("eval() argument 1 must be string or code object");
+
+            string strExpression;
+            switch (expression) {
```

Comment:
Thanks for the comment, I like the overload approach much better! I simply copied the pattern from `compile` which I thought was what it supposed to be. But `compile` can accept more types than just `str`, `bytes`, and `bytearray` (e.g.  `_ast.AST)`. That's why its signature has `object`. However, I can split `compile` up into overloads as well, if you think it is a good idea.

> assuming the bytes are UTF-8 (which is what you seem to indicate CPython does).

This is not exactly correct. CPython 3 assumes UTF-8 *by default*, just as CPython 2 assumes Latin 1 by default. But, as per [PEP 263](https://www.python.org/dev/peps/pep-0263/), the encoding can be specified explicitly in the first two lines of code and can be any encoding, including wide char encodings like UTF-16. So for now UTF-8 is good enough, but eventually the encoding has to be autodetected. This behavior is in Python since 2.3, but I don't know if IronPython does it already, I haven't found it yet.

Reply:
Hmm, does PEP263 cover byte inputs to eval/exec/compile or just source files?

Reply:
> does PEP263 cover byte inputs to eval/exec/compile or just source files?

PEP 263 does not mention eval/exec/compile except for a small note in the history log that one implementation step required tweaks in compile. However, when you look in [test_source_encoding.py](https://github.com/IronLanguages/ironpython3/blob/master/Src/StdLib/Lib/test/test_source_encoding.py) you can notice that it makes heavy use of exec/compile to test PEP 263 and PEP 3120 cases. So my conclusion is that eval/exec/compile should treat their input as source files would be teated.

Reply:
> eval/exec/compile should treat ...

Not sure about `eval`. Will need to do some more experiments or read CPython source code. Also, all examples of `compile` use `exec` mode. Maybe in other modes it is not so.

Reply:
Ok, looks like `eval` is also affected by this since `eval(b'#coding: ascii\n"\xff"')` gives a `SyntaxError`. Note that the string version (`exec('#coding: ascii\n"\xff"')`) works so I wonder if it only checks the coding for bytes?

You could try checking the coding, but I think decoding to UTF8 is likely more than enough to cover usage outside of testing.

Reply:
> decoding to UTF8 is likely more than enough

I totally agree. This is what I am aiming at in this PR. If time permits, I will do checks for explicit coding, but there are other more important missing bits in Python 3 right now. The whole explicit coding syntax in a comment feels like a legacy workaround from the past. Nowadays, if anyone wants to use characters from outside ASCII in their source file, UTF-8 is a prevalent and widely supported option.

Reply:
@slozier You mark your example method with `[LightThrowing]` attribute, what does it mean? More specifically, when it should be used and when not? I am looking at the `eval` method: some of the overloads have `[LightThrowing]` and some don't. There seem to be no clear way to tell a fundamental difference between them. They all throw `PythonOps.TypeError`.

Reply:
I'm not too familiar with it, but the way I understand it is it's a mechanism to throw without throwing. If a method is `LightThrowing` it means that it may return a `LightException` in which case some magic (I have no idea how this part works) will happen to make it throw. In this case it looks like `code.Call(scope)` may return a `LightException` so you need to tag the method.

Reply:
>  looks like `code.Call(scope)` may return a `LightException`

It still does not make sense to me: in such case the other overload should have been tagged as well:
```Python
 public static object eval(CodeContext/*!*/ context, FunctionCode code, ...)
```
But `object Call(CodeContext/*!*/ context)` is not tagged with `[LightThrowing]` nor do I see it returning a light exception anywhere.  Frankly, nowhere I see `LightException.Throw` used in the entire solution.

Reply:
Another interesting issue: as a consequence of using overloads instead of the `is`/`as`/`if` machinery, error reporting changes (for worse). Consider this:
```python
eval(1)
```
Python 3.4: `eval() arg 1 must be a string, bytes or code object`
IronPython 3: `expected code, got int` (because `code` is the first overload type, the other two are not mentioned)

It can be fixed by adding a catch-all overload with `object` parameter type, which then reports a correct type, but I think it would be better to fix it at the dispatcher level, for all such cases at once.

A similar case, but somewhat different:
```python
eval(None)
```
Python 3.4: `eval() arg 1 must be a string, bytes or code object`
IronPython 3: `Multiple targets could match: eval(code, dict, object), eval(IList[Byte], dict, object), eval(str, dict, object)`

Not sure where or how this one could be fixed, or even if we care.

Reply:
The `eval` overload you point out does not calling into the `[LightThrowing] eval` overloads and so does not need to be tagged (although I have no idea if a user can somehow generate a `FunctionCode` object which is `LightThrowing` in which case calling it might be bugged?)

You only need to tag the public API (e.g. things you call from Python). Since `Call` is internal it does not need to be tagged.

I'm not familiar with how all this works (although I'm learning stuff be responding to this), but I find it's best to go by example. So go find the `IronPython.Modules.Builtin.chr` method, comment out `[LightThrowing]` then try to execute `chr(0x200000)` and you'll see that instead of throwing it'll return a `LightException` object.

Reply:
The `expected code, got int` instead of `eval() arg 1 must be a string, bytes or code object` issue is quite common. There's probably something we could do to check all the overloads and come up with a better error message. Personally I don't really care about it (at this point in time) as long as it throws the proper exception type. 

As for the `Multiple targets could match` issue, you can c=tag the first argument with `[NotNull]`.

Reply:
Got it, works like a charm.